contracts/validator/contract/XDCValidatorV2.sol

```diff
@@ -0,0 +1,518 @@
+    }
+
+    function resign(
+        address _candidate
+    ) public onlyOwner(_candidate) onlyCandidate(_candidate) {
+        validatorsState[_candidate].isCandidate = false;
+        candidateCount = candidateCount.sub(1);
+
+        deleteCandidate(_candidate);
+
+        // Cleanup the ownerToCandidate mapping for the resigning candidate's owner
+        address[] storage ownedCandidates = ownerToCandidate[msg.sender];
+        uint256 ownedCandidatesLength = ownedCandidates.length;
+        for (uint256 j = 0; j < ownedCandidatesLength; j++) {
+            if (ownedCandidates[j] == _candidate) {
+                ownedCandidates[j] = ownedCandidates[ownedCandidatesLength - 1];
+                delete ownedCandidates[ownedCandidatesLength - 1];
+                ownedCandidates.length--; // Manually decrease the array length
+                break;
+            }
+        }
+
+        // Optionally, consider adjusting ownerCount if needed
+        if (ownedCandidates.length == 0) {
+            // If specific logic is needed to manage the owners array, implement here
+            deleteOwner(msg.sender);
+        }
+
+        uint256 cap = validatorsState[_candidate].voters[msg.sender];
+        validatorsState[_candidate].cap = validatorsState[_candidate].cap.sub(
+            cap
+        );
+        validatorsState[_candidate].voters[msg.sender] = 0;
+        // refunding after resigning X blocks
+        uint256 withdrawBlockNumber = candidateWithdrawDelay.add(block.number);
+        withdrawsState[msg.sender].caps[withdrawBlockNumber] = withdrawsState[
+            msg.sender
+        ].caps[withdrawBlockNumber].add(cap);
+        withdrawsState[msg.sender].blockNumbers.push(withdrawBlockNumber);
+        emit Resign(msg.sender, _candidate);
+    }
+
+    // voteInvalidKYC : any candidate can vote for invalid KYC i.e. a particular candidate's owner has uploaded a bad KYC.
+    // On securing 75% votes against an owner ( not candidate ), owner & all its candidates will lose their funds.
+    function voteInvalidKYC(
+        address _owner
+    ) public onlyValidCandidate(msg.sender) {
+        address candidateOwner = getCandidateOwner(msg.sender);
+
+        require(!hasVotedInvalid[candidateOwner][_owner], "Already voted");
+        hasVotedInvalid[candidateOwner][_owner] = true;
+        invalidKYCCount[_owner] += 1;
+        if ((invalidKYCCount[_owner] * 100) / getOwnerCount() >= 75) {
+            // 75% owners say that the KYC is invalid
+            address[] memory allMasternodes = new address[](
+                candidates.length - 1
+            );
+            uint count = 0;
+            uint j = 0;
+            address[] memory newCandidates = new address[](candidates.length);
+            invalidOwner[_owner] = true;
+            for (uint i = 0; i < candidates.length; i++) {
+                if (getCandidateOwner(candidates[i]) == _owner) {
+                    // logic to remove cap.
```

Comment:
consider cache the element candidates[i] or entire array candidates to save gas

Reply:
fixed